pkg/controllers/scheduling/scheduling_controller.go

```diff
@@ -42,31 +46,61 @@ type schedulingController struct {
 	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister
 	placementLister         clusterlisterv1alpha1.PlacementLister
 	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister
+	enqueuePlacementFunc    enqueuePlacementFunc
 	scheduleFunc            scheduleFunc
 }
 
 // NewDecisionSchedulingController return an instance of schedulingController
 func NewSchedulingController(
 	clusterClient clusterclient.Interface,
-	clusterLister clusterlisterv1.ManagedClusterLister,
-	clusterSetLister clusterlisterv1alpha1.ManagedClusterSetLister,
-	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister,
+	clusterInformer clusterinformerv1.ManagedClusterInformer,
+	clusterSetInformer clusterinformerv1alpha1.ManagedClusterSetInformer,
+	clusterSetBindingInformer clusterinformerv1alpha1.ManagedClusterSetBindingInformer,
 	placementInformer clusterinformerv1alpha1.PlacementInformer,
 	placementDecisionInformer clusterinformerv1alpha1.PlacementDecisionInformer,
 	recorder events.Recorder,
 ) factory.Controller {
+	syncCtx := factory.NewSyncContext(schedulingControllerName, recorder)
+	enqueuePlacementFunc := func(namespace, name string) {
+		syncCtx.Queue().Add(fmt.Sprintf("%s/%s", namespace, name))
+	}
+
 	// build controller
 	c := schedulingController{
 		clusterClient:           clusterClient,
-		clusterLister:           clusterLister,
-		clusterSetLister:        clusterSetLister,
-		clusterSetBindingLister: clusterSetBindingLister,
+		clusterLister:           clusterInformer.Lister(),
+		clusterSetLister:        clusterSetInformer.Lister(),
+		clusterSetBindingLister: clusterSetBindingInformer.Lister(),
 		placementLister:         placementInformer.Lister(),
 		placementDecisionLister: placementDecisionInformer.Lister(),
 		scheduleFunc:            schedule,
+		enqueuePlacementFunc:    enqueuePlacementFunc,
 	}
 
+	// setup event handler for cluster informer
+	clusterInformer.Informer().AddEventHandler(&clusterEventHandler{
```

Comment:
I'm thinking about what I think these event handlers will do before the informers themselves are sync'd.   Since they are trying to use the listers, I think the results will be inconsistent.  However, I don't think the sync itself will complete if you wait.

Since every placement will be queue when you first begin, not adding something to the list is ok.

If your reasoning is the same, please indicate in the individual eventhandler structs the limitations and reference the comment here.  I'll check back tomorrow morning.

Reply:
Yes, the reason is the same. Comment added as below.
>// Once a cluster changes, clusterEventHandler enqueues all placements which are
	// impacted potentially for further reconciliation. It might not function before the
	// informers/listers of clusterset/clustersetbinding/placement are synced during
	// controller booting. But that should not cause any problem because all existing
	// placements will be enqueued by the controller anyway when booting.